File: exp-09/q2-n-queens/solution.go
package main

import "fmt"

const max = 100

var (
	n     int
	count int
	board [max]int
)

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func isSafe(row, col int) bool {
	for i := 0; i < row; i++ {
		if board[i] == col || abs(board[i]-col) == abs(i-row) {
			return false
		}
	}
	return true
}

func solve(row int) {
	if row == n {
		for i := 0; i < n; i++ {
			if i > 0 {
				fmt.Print(" ")
			}
			fmt.Print(board[i] + 1)
		}
		fmt.Println()
		count++
		return
	}
	for col := 0; col < n; col++ {
		if isSafe(row, col) {
			board[row] = col
			solve(row + 1)
		}
	}
}

func main() {
	fmt.Scan(&n)
	solve(0)
	fmt.Printf("Total solutions: %d\n", count)
}

File: exp-09/q2-n-queens/Solution.java
import java.util.Scanner;

public class Solution {
    private static final int MAX = 100;
    private static int n;
    private static int count = 0;
    private static int[] board = new int[MAX];

    static boolean isSafe(int row, int col) {
        for (int i = 0; i < row; i++) {
            if (board[i] == col || Math.abs(board[i] - col) == Math.abs(i - row))
                return false;
        }
        return true;
    }

    static void solve(int row) {
        if (row == n) {
            for (int i = 0; i < n; i++) {
                if (i > 0)
                    System.out.print(" ");
                System.out.print(board[i] + 1);
            }
            System.out.println();
            count++;
            return;
        }
        for (int col = 0; col < n; col++) {
            if (isSafe(row, col)) {
                board[row] = col;
                solve(row + 1);
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        solve(0);
        System.out.println("Total solutions: " + count);
        sc.close();
    }
}
